proposed/http-client/http-client.md

```diff
@@ -7,43 +7,64 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
-MUST throw a `Psr\Http\Client\RequestExceptionInterface`. If there is an error
-with the network or the remote server cannot be reached, the HTTP client MUST throw
-a `Psr\Http\Client\NetworkExceptionInterface`.
+## Client
 
-Smaller issues that do not block the client from sending the request (such as
-invalid HTTP versions) MUST NOT result in exceptions.
+A Client is an object implementing `ClientInterface`.
 
-If the remote server answers with a response that can be parsed into a PSR-7 response,
-the client MUST NOT throw an exception. For example, response status codes in the
-400 and 500 range MUST NOT cause an exception.
+A Client MAY:
 
-## Goal
+* Elect to send an altered HTTP request from the one it was provided.  For example, it could
+compress an outgoing message body.
+* Elect to alter a received HTTP response before returning it to the calling library. For example, it could
+decompress an incoming message body.
 
-The goal of this PSR is to allow developers to create libraries decoupled from HTTP client
-implementations. This will make libraries more reusable as it reduces the number of
-dependencies and lowers the likelihood of version conflicts.
+If a Client chooses to alter either the HTTP request or HTTP response, it MUST ensure that the
+response remains internally consistent.  For example, if a Client chooses to decompress the message
+body then it MUST also remove the `Content-Encoding` header and adjust the `Content-Length` header.
 
-A second goal is that HTTP clients can be replaced as per the
-[Liskov substitutions principle][Liskov]. This means that all clients MUST behave in the
-same way when sending a request.
+Note that as a result, since [PSR-7 objects are immutable](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-7-http-message-meta.md#why-value-objects),
+the Calling Library MUST NOT assume that the object passed to `ClientInterface::sendRequest()` will be the same PHP object
+that is actually sent. For example, the Request object that is returned by an exception MAY be a different object than
+the one passed to `sendRequest()`, so comparison by reference (===) is not possible.
+
+A Client MUST:
+
+* Reassemble a multi-step HTTP 1xx response itself so that what is returned to the Calling Library is an valid HTTP response
+of status code 200 or higher.
+
+## Error handling
+
+A Client MUST NOT treat a well-formed HTTP request or HTTP response as an error condition.  For example, response
+status codes in the 400 and 500 range MUST NOT cause an exception and MUST be returned to the Calling Library as normal.
+
+A Client MUST throw an instance of `Psr\Http\Client\ClientExceptionInterface` if and only if it is unable to send
+the HTTP request at all or if no HTTP response is ever returned (such as a timeout).
+
+If a request cannot be sent because the request message is not a valid HTTP request, the Client MUST throw an instance
+of `Psr\Http\Client\RequestExceptionInterface`.
```

Comment:
Just note. Sending a Request with HTTP version 5.0 is not invalid. Should we mention something like that here?

Reply:
That's a can of worms because then you get into defining what constitutes "valid".  Perhaps the slightly less restrictive "message is malformed"?  (I informally get the intent here, but specifying it formally is tricky.)

Reply:
Exactly. Messages that are not malformed is a good expression. 

Like messsges without a URL or host is impossible to send.